crates/core/tedge_agent/src/entity_manager/server.rs

```diff
@@ -307,23 +313,40 @@ impl EntityStoreServer {
     async fn deregister_entity(&mut self, topic_id: &EntityTopicId) -> Vec<EntityMetadata> {
         let deleted = self.entity_store.deregister_entity(topic_id);
         for entity in deleted.iter().rev() {
-            for twin_key in entity.twin_data.keys() {
-                let topic = self.mqtt_schema.topic_for(
-                    &entity.topic_id,
-                    &Channel::EntityTwinData {
-                        fragment_key: twin_key.to_string(),
-                    },
-                );
-                let clear_twin_msg = MqttMessage::new(&topic, "").with_retain();
-
-                self.publish_message(clear_twin_msg).await;
+            // Clear all retained metadata/data associated with the entity
+            match self.mqtt_connector.connection_for(&entity.topic_id).await {
+                Ok(_) => {
+                    if let Some(messages) =
+                        self.mqtt_connector.get_message_receiver(&entity.topic_id)
+                    {
+                        while let Ok(Some(message)) =
+                            timeout(Duration::from_secs(1), messages.next()).await
+                        {
+                            if message.retain && !message.payload_bytes().is_empty() {
+                                let clear_msg = MqttMessage::new(&message.topic, "").with_retain();
+                                if let Err(err) = self.mqtt_publisher.send(clear_msg).await {
+                                    error!(
+                                    "Failed to clear retained message on topic {} while de-registering {} due to: {err}",
+                                    topic_id,
+                                    message.topic
+                                );
+                                }
+                            }
+                        }
+                    }
+
+                    self.mqtt_connector.disconnect(&entity.topic_id).await;
```

Comment:
Why does `disconnect` need to known the topic id (or filter). Is the MQTT connection shared among all the virtual connection for a topic id? Wouldn't be simpler to have a fresh MQTT connection each time an entity is deregistered (something that really only rarely happens - making no point is caching/sharing resources)?

Reply:
Resolved by 67a05b3